File: cpp/benchmarks/mixed_lifecycle.cpp
#if defined(LLAB_NATIVE_V1)
#include "llab/order_book.hpp"
#elif defined(LLAB_PARITY_V2)
#include "llab/parity_order_book.hpp"
#elif defined(LLAB_DENSE_V3)
#include "llab/dense_ladder_order_book.hpp"
#elif defined(LLAB_BITMAP_V4)
#include "llab/bitmap_ladder_order_book.hpp"
#elif defined(LLAB_BITMAP_V5)
#include "llab/bitmap_backshift_order_book.hpp"
#elif defined(LLAB_BITMAP_V6)
#include "llab/bitmap_packed_order_book.hpp"
#else
#error "select one order-book version"
#endif

#if defined(LLAB_METRICS_DISABLED) || defined(LLAB_METRICS_CAPTURE) ||                             \
    defined(LLAB_METRICS_SAMPLED)
#include "llab/metrics.hpp"
#endif

#include <algorithm>
#include <chrono>
#include <cmath>
#include <cstdint>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <stdexcept>
#include <string>
#include <string_view>
#include <vector>

namespace {

#if defined(LLAB_NATIVE_V1)
using namespace llab::order_book;
constexpr const char *kVersion = "native_v1";
#elif defined(LLAB_PARITY_V2)
using namespace llab::parity_order_book;
constexpr const char *kVersion = "parity_v2";
#elif defined(LLAB_DENSE_V3)
using namespace llab::dense_ladder_order_book;
constexpr const char *kVersion = "dense_ladder_v3";
#elif defined(LLAB_BITMAP_V4)
using namespace llab::bitmap_ladder_order_book;
constexpr const char *kVersion = "bitmap_ladder_v4";
#elif defined(LLAB_BITMAP_V5)
using namespace llab::bitmap_backshift_order_book;
constexpr const char *kVersion = "bitmap_backshift_v5";
#else
using namespace llab::bitmap_packed_order_book;
constexpr const char *kVersion = "bitmap_packed_v6";
#endif

constexpr std::size_t kCycles = 8'192;
constexpr std::size_t kMaximumLiveOrders = 32'768;
constexpr std::size_t kWarmupSamples = 10;
constexpr std::size_t kMeasuredSamples = 200;

OrderBook make_book() {
#if defined(LLAB_DENSE_V3) || defined(LLAB_BITMAP_V4) || defined(LLAB_BITMAP_V5) ||                \
    defined(LLAB_BITMAP_V6)
    return OrderBook(kMaximumLiveOrders, 99'873, 256);
#else
    return OrderBook(kMaximumLiveOrders);
#endif
}

std::vector<Event> make_trace() {
    std::vector<Event> events;
    events.reserve(kCycles * 7);
    for (std::size_t cycle = 0; cycle < kCycles; ++cycle) {
        const OrderId a = cycle * 3 + 1;
        const OrderId b = cycle * 3 + 2;
        const OrderId c = cycle * 3 + 3;
        const Price offset = cycle % 128;
        const Side a_side = cycle % 2 == 0 ? Side::Bid : Side::Ask;
        const Side b_side = a_side == Side::Bid ? Side::Ask : Side::Bid;
        const Price a_price = a_side == Side::Bid ? 100'000 - offset : 100'001 + offset;
        const Price b_price = b_side == Side::Bid ? 100'000 - offset : 100'001 + offset;

        events.push_back({EventType::Add, a, 0, a_side, a_price, 10});
        events.push_back({EventType::Add, b, 0, b_side, b_price, 20});
        events.push_back({EventType::Cancel, a, 0, a_side, 0, 3});
        events.push_back({EventType::OrderUpsert, b, 0, b_side, b_price, 15});
        events.push_back({EventType::Replace, a, c, a_side, a_price, 7});
        events.push_back({EventType::Execute, c, 0, a_side, 0, 7});
        events.push_back({cycle % 2 == 0 ? EventType::Delete : EventType::OrderDelete, b});
    }
    return events;
}

std::uint64_t fold_bbo(const OrderBook &book) {
    std::uint64_t digest = 0x9E3779B97F4A7C15ULL;
    if (const auto bid = book.best_bid())
        digest ^= bid->price_ticks ^ (bid->quantity << 1U);
    if (const auto ask = book.best_ask())
        digest ^= (ask->price_ticks << 7U) ^ (ask->quantity << 3U);
    return digest;
}

std::uint64_t run_sample(const std::vector<Event> &events, std::uint64_t &digest) {
    auto book = make_book();
#if defined(LLAB_METRICS_CAPTURE)
    llab::metrics::ThreadRing metric_ring(events.size());
#elif defined(LLAB_METRICS_SAMPLED)
    constexpr std::size_t metric_sample_mask = 63;
    llab::metrics::ThreadRing metric_ring((events.size() + metric_sample_mask) /
                                          (metric_sample_mask + 1));
    std::size_t metric_sequence = 0;
#endif
    const auto start = std::chrono::steady_clock::now();
    for (const Event &event : events) {
#if defined(LLAB_METRICS_DISABLED)
        llab::metrics::Scope<false> metric_scope(nullptr, 0);
        book.apply(event);
#elif defined(LLAB_METRICS_SAMPLED)
        if ((metric_sequence++ & metric_sample_mask) == 0) {
            llab::metrics::Scope<true> metric_scope(&metric_ring, 1);
            book.apply(event);
        } else {
            book.apply(event);
        }
#elif defined(LLAB_METRICS_CAPTURE)
        llab::metrics::Scope<true> metric_scope(&metric_ring, 1);
        book.apply(event);
#else
        book.apply(event);
#endif
#if defined(LLAB_READ_BBO)
        digest ^= fold_bbo(book) + 0x9E3779B97F4A7C15ULL + (digest << 6U) + (digest >> 2U);
#endif
    }
    const auto end = std::chrono::steady_clock::now();
#if defined(LLAB_METRICS_CAPTURE)
    if (metric_ring.used() != events.size() || metric_ring.dropped() != 0)
        throw std::runtime_error("metrics capture incomplete");
#elif defined(LLAB_METRICS_SAMPLED)
    if (metric_ring.used() != (events.size() + metric_sample_mask) / (metric_sample_mask + 1) ||
        metric_ring.dropped() != 0)
        throw std::runtime_error("metrics sampled capture incomplete");
#endif
    if (book.live_order_count() != 0 || book.best_bid() || book.best_ask())
        throw std::runtime_error("mixed lifecycle did not restore empty book");
    return std::chrono::duration_cast<std::chrono::nanoseconds>(end - start).count();
}

void verify_trace(const std::vector<Event> &events) {
    auto book = make_book();
    for (const Event &event : events)
        book.apply(event);
    if (book.live_order_count() != 0 || book.best_bid() || book.best_ask())
        throw std::runtime_error("mixed lifecycle did not restore empty book");
}

std::size_t rank(std::size_t count, long double percentile) {
    return static_cast<std::size_t>(std::ceil(percentile * count)) - 1;
}
} // namespace

int main(int argc, char **argv) {
    std::string raw_path;
    bool verify_only = false;
    if (argc == 3 && std::string_view(argv[1]) == "--raw")
        raw_path = argv[2];
    else if (argc == 2 && std::string_view(argv[1]) == "--verify")
        verify_only = true;
    else if (argc != 1) {
        std::cerr << "usage: mixed_lifecycle [--raw PATH | --verify]\n";
        return 2;
    }

    const auto events = make_trace();
    if (verify_only) {
        verify_trace(events);
        return 0;
    }
    const auto layout_book = make_book();
    std::uint64_t digest = 0;
    for (std::size_t sample = 0; sample < kWarmupSamples; ++sample)
        (void)run_sample(events, digest);
    std::vector<std::uint64_t> samples;
    samples.reserve(kMeasuredSamples);
    for (std::size_t sample = 0; sample < kMeasuredSamples; ++sample)
        samples.push_back(run_sample(events, digest));
    if (!raw_path.empty()) {
        std::ofstream raw(raw_path);
        if (!raw)
            throw std::runtime_error("cannot write raw samples");
        raw << "sample_index,duration_ns\n";
        for (std::size_t index = 0; index < samples.size(); ++index)
            raw << index << ',' << samples[index] << '\n';
    }

    auto sorted = samples;
    std::sort(sorted.begin(), sorted.end());
    long double total = 0;
    for (const auto sample : samples)
        total += sample;
    const auto p50 = sorted[rank(sorted.size(), 0.50L)];
    std::cout << std::fixed << std::setprecision(2);
    std::cout << "format: llab_benchmark_v1\nbenchmark: stream_mixed_lifecycle_" << kVersion
#if defined(LLAB_READ_BBO)
              << "_with_bbo_v1\n";
#else
              << "_v1\n";
#endif
    std::cout << "language: cpp\nevent_size_bytes: " << sizeof(Event)
              << "\norder_book_size_bytes: " << sizeof(OrderBook)
              << "\norder_index_capacity: " << layout_book.order_index_capacity()
              << "\nevents_per_sample: " << events.size() << "\nwarmup_samples: " << kWarmupSamples
              << "\nmeasured_samples: " << kMeasuredSamples << "\nmin_ns: " << sorted.front()
              << "\np50_ns: " << p50 << "\np90_ns: " << sorted[rank(sorted.size(), 0.90L)]
              << "\np99_ns: " << sorted[rank(sorted.size(), 0.99L)] << "\nmax_ns: " << sorted.back()
              << "\nmean_ns: " << total / samples.size()
              << "\np50_ns_per_event: " << static_cast<long double>(p50) / events.size()
              << "\np50_events_per_s: "
              << static_cast<long double>(events.size()) * 1'000'000'000.0L / p50
              << "\nresult_digest: " << digest << '\n';
}

File: rust/src/bin/mixed_lifecycle.rs
use low_latency_lab_benchmarks::bitmap_backshift_order_book;
use low_latency_lab_benchmarks::bitmap_ladder_order_book;
use low_latency_lab_benchmarks::bitmap_packed_order_book;
use low_latency_lab_benchmarks::dense_ladder_order_book;
use low_latency_lab_benchmarks::order_book;
use low_latency_lab_benchmarks::parity_order_book;
use std::{fs::File, io::Write, time::Instant};

const CYCLES: usize = 8_192;
const MAXIMUM_LIVE_ORDERS: usize = 32_768;
const WARMUP_SAMPLES: usize = 10;
const MEASURED_SAMPLES: usize = 200;

#[derive(Clone, Copy)]
struct LogicalEvent {
    kind: parity_order_book::EventType,
    order_id: u64,
    replacement_order_id: u64,
    side: parity_order_book::Side,
    price_ticks: u64,
    quantity: u64,
}

fn trace() -> Vec<LogicalEvent> {
    let mut events = Vec::with_capacity(CYCLES * 7);
    for cycle in 0..CYCLES {
        let a = cycle as u64 * 3 + 1;
        let b = a + 1;
        let c = a + 2;
        let offset = cycle as u64 % 128;
        let a_side = if cycle % 2 == 0 {
            parity_order_book::Side::Bid
        } else {
            parity_order_book::Side::Ask
        };
        let b_side = if a_side == parity_order_book::Side::Bid {
            parity_order_book::Side::Ask
        } else {
            parity_order_book::Side::Bid
        };
        let price = |side| {
            if side == parity_order_book::Side::Bid { 100_000 - offset } else { 100_001 + offset }
        };
        events.extend([
            LogicalEvent {
                kind: parity_order_book::EventType::Add,
                order_id: a,
                replacement_order_id: 0,
                side: a_side,
                price_ticks: price(a_side),
                quantity: 10,
            },
            LogicalEvent {
                kind: parity_order_book::EventType::Add,
                order_id: b,
                replacement_order_id: 0,
                side: b_side,
                price_ticks: price(b_side),
                quantity: 20,
            },
            LogicalEvent {
                kind: parity_order_book::EventType::Cancel,
                order_id: a,
                replacement_order_id: 0,
                side: a_side,
                price_ticks: 0,
                quantity: 3,
            },
            LogicalEvent {
                kind: parity_order_book::EventType::OrderUpsert,
                order_id: b,
                replacement_order_id: 0,
                side: b_side,
                price_ticks: price(b_side),
                quantity: 15,
            },
            LogicalEvent {
                kind: parity_order_book::EventType::Replace,
                order_id: a,
                replacement_order_id: c,
                side: a_side,
                price_ticks: price(a_side),
                quantity: 7,
            },
            LogicalEvent {
                kind: parity_order_book::EventType::Execute,
                order_id: c,
                replacement_order_id: 0,
                side: a_side,
                price_ticks: 0,
                quantity: 7,
            },
            LogicalEvent {
                kind: if cycle % 2 == 0 {
                    parity_order_book::EventType::Delete
                } else {
                    parity_order_book::EventType::OrderDelete
                },
                order_id: b,
                replacement_order_id: 0,
                side: b_side,
                price_ticks: 0,
                quantity: 0,
            },
        ]);
    }
    events
}

fn run_native(events: &[LogicalEvent], read_bbo: bool, digest: &mut u64) -> u128 {
    let mut book = order_book::OrderBook::with_order_capacity(MAXIMUM_LIVE_ORDERS);
    let start = Instant::now();
    for e in events {
        use order_book::{Event, Side};
        let side = if e.side == parity_order_book::Side::Bid { Side::Bid } else { Side::Ask };
        let event = match e.kind {
            parity_order_book::EventType::Add => Event::Add {
                order_id: e.order_id,
                side,
                price_ticks: e.price_ticks,
                quantity: e.quantity,
            },
            parity_order_book::EventType::Cancel => {
                Event::Cancel { order_id: e.order_id, quantity: e.quantity }
            }
            parity_order_book::EventType::Execute => {
                Event::Execute { order_id: e.order_id, quantity: e.quantity }
            }
            parity_order_book::EventType::Delete => Event::Delete { order_id: e.order_id },
            parity_order_book::EventType::Replace => Event::Replace {
                old_order_id: e.order_id,
                new_order_id: e.replacement_order_id,
                side,
                price_ticks: e.price_ticks,
                quantity: e.quantity,
            },
            parity_order_book::EventType::OrderUpsert => Event::OrderUpsert {
                order_id: e.order_id,
                side,
                price_ticks: e.price_ticks,
                quantity: e.quantity,
            },
            parity_order_book::EventType::OrderDelete => {
                Event::OrderDelete { order_id: e.order_id }
            }
        };
        book.apply(event).unwrap();
        if read_bbo {
            let mut value = 0x9E37_79B9_7F4A_7C15;
            if let Some(level) = book.best_bid() {
                value ^= level.price_ticks ^ (level.quantity << 1);
            }
            if let Some(level) = book.best_ask() {
                value ^= (level.price_ticks << 7) ^ (level.quantity << 3);
            }
            *digest ^= value
                .wrapping_add(0x9E37_79B9_7F4A_7C15)
                .wrapping_add(*digest << 6)
                .wrapping_add(*digest >> 2);
        }
    }
    let elapsed = start.elapsed().as_nanos();
    assert_eq!(book.live_order_count(), 0);
    assert_eq!(book.best_bid(), None);
    assert_eq!(book.best_ask(), None);
    elapsed
}

fn run_parity(
    events: &[LogicalEvent],
    dense: bool,
    bitmap: bool,
    backshift: bool,
    packed: bool,
    read_bbo: bool,
    digest: &mut u64,
) -> u128 {
    if dense {
        let mut book = dense_ladder_order_book::OrderBook::new(MAXIMUM_LIVE_ORDERS, 99_873, 256);
        let start = Instant::now();
        for &e in events {
            book.apply(parity_event(e)).unwrap();
            if read_bbo {
                *digest ^=
                    fold_parity_bbo(book.best_bid(), book.best_ask()).wrapping_add(*digest << 1);
            }
        }
        assert_eq!(book.live_order_count(), 0);
        assert_eq!(book.best_bid(), None);
        assert_eq!(book.best_ask(), None);
        return start.elapsed().as_nanos();
    } else if bitmap {
        let mut book = bitmap_ladder_order_book::OrderBook::new(MAXIMUM_LIVE_ORDERS, 99_873, 256);
        let start = Instant::now();
        for &e in events {
            book.apply(parity_event(e)).unwrap();
            if read_bbo {
                *digest ^=
                    fold_parity_bbo(book.best_bid(), book.best_ask()).wrapping_add(*digest << 1);
            }
        }
        assert_eq!(book.live_order_count(), 0);
        assert_eq!(book.best_bid(), None);
        assert_eq!(book.best_ask(), None);
        return start.elapsed().as_nanos();
    } else if backshift {
        let mut book =
            bitmap_backshift_order_book::OrderBook::new(MAXIMUM_LIVE_ORDERS, 99_873, 256);
        let start = Instant::now();
        for &e in events {
            book.apply(parity_event(e)).unwrap();
            if read_bbo {
                *digest ^=
                    fold_parity_bbo(book.best_bid(), book.best_ask()).wrapping_add(*digest << 1);
            }
        }
        assert_eq!(book.live_order_count(), 0);
        assert_eq!(book.best_bid(), None);
        assert_eq!(book.best_ask(), None);
        return start.elapsed().as_nanos();
    } else if packed {
        let mut book = bitmap_packed_order_book::OrderBook::new(MAXIMUM_LIVE_ORDERS, 99_873, 256);
        let start = Instant::now();
        for &e in events {
            book.apply(parity_event(e)).unwrap();
            if read_bbo {
                *digest ^=
                    fold_parity_bbo(book.best_bid(), book.best_ask()).wrapping_add(*digest << 1);
            }
        }
        assert_eq!(book.live_order_count(), 0);
        assert_eq!(book.best_bid(), None);
        assert_eq!(book.best_ask(), None);
        return start.elapsed().as_nanos();
    } else {
        let mut book = parity_order_book::OrderBook::new(MAXIMUM_LIVE_ORDERS);
        let start = Instant::now();
        for &e in events {
            book.apply(parity_event(e)).unwrap();
            if read_bbo {
                *digest ^=
                    fold_parity_bbo(book.best_bid(), book.best_ask()).wrapping_add(*digest << 1);
            }
        }
        assert_eq!(book.live_order_count(), 0);
        assert_eq!(book.best_bid(), None);
        assert_eq!(book.best_ask(), None);
        return start.elapsed().as_nanos();
    }
}

fn fold_parity_bbo(
    bid: Option<parity_order_book::Level>,
    ask: Option<parity_order_book::Level>,
) -> u64 {
    let mut value = 0x9E37_79B9_7F4A_7C15;
    if let Some(level) = bid {
        value ^= level.price_ticks ^ (level.quantity << 1);
    }
    if let Some(level) = ask {
        value ^= (level.price_ticks << 7) ^ (level.quantity << 3);
    }
    value
}

fn parity_event(e: LogicalEvent) -> parity_order_book::Event {
    parity_order_book::Event {
        order_id: e.order_id,
        replacement_order_id: e.replacement_order_id,
        price_ticks: e.price_ticks,
        quantity: e.quantity,
        event_type: e.kind,
        side: e.side,
    }
}

pub fn main() {
    let name = option_env!("CARGO_BIN_NAME").unwrap();
    let raw_path = match std::env::args().collect::<Vec<_>>().as_slice() {
        [_] => None,
        [_, flag, path] if flag == "--raw" => Some(path.clone()),
        _ => panic!("usage: mixed_lifecycle [--raw PATH]"),
    };
    let read_bbo = name.ends_with("_with_bbo");
    let variant = name.trim_end_matches("_with_bbo");
    let events = trace();
    let mut digest = 0;
    let mut run = || match variant {
        "native_mixed_lifecycle" => run_native(&events, read_bbo, &mut digest),
        "parity_mixed_lifecycle" => {
            run_parity(&events, false, false, false, false, read_bbo, &mut digest)
        }
        "dense_mixed_lifecycle" => {
            run_parity(&events, true, false, false, false, read_bbo, &mut digest)
        }
        "bitmap_mixed_lifecycle" => {
            run_parity(&events, false, true, false, false, read_bbo, &mut digest)
        }
        "bitmap_backshift_mixed_lifecycle" => {
            run_parity(&events, false, false, true, false, read_bbo, &mut digest)
        }
        "bitmap_packed_mixed_lifecycle" => {
            run_parity(&events, false, false, false, true, read_bbo, &mut digest)
        }
        _ => panic!("unknown binary"),
    };
    for _ in 0..WARMUP_SAMPLES {
        let _ = run();
    }
    let mut samples = Vec::with_capacity(MEASURED_SAMPLES);
    for _ in 0..MEASURED_SAMPLES {
        samples.push(run());
    }
    if let Some(path) = raw_path {
        let mut raw = File::create(path).expect("cannot write raw samples");
        writeln!(raw, "sample_index,duration_ns").unwrap();
        for (index, sample) in samples.iter().enumerate() {
            writeln!(raw, "{index},{sample}").unwrap();
        }
    }
    samples.sort_unstable();
    println!(
        "format: llab_benchmark_v1\nbenchmark: stream_mixed_lifecycle_{name}_v1\nlanguage: rust\nevents_per_sample: {}\nwarmup_samples: {WARMUP_SAMPLES}\nmeasured_samples: {MEASURED_SAMPLES}\np50_ns: {}\np99_ns: {}\nresult_digest: {digest}",
        events.len(),
        samples[99],
        samples[197]
    );
}
